### PR TITLE
Implement Lod range

### PR DESCRIPTION
**Pull request by Windfisch**
This implements the lod_max_distance / lod_min_distance and hysteresis of `GeometryInstance`

### CODE REVIEW

File: doc/classes/GeometryInstance.xml [hunk 1]
```diff
@@ -24,12 +24,14 @@
 		</member>
 		<member name="lod_max_hysteresis" type="float" setter="set_lod_max_hysteresis" getter="get_lod_max_hysteresis">
 			The GeometryInstance's max LOD margin.
+			Margins are symmetrical around lod_max_distance.
```
Comment (akien-mga) [inline doc/classes/GeometryInstance.xml]: To make a proper hyperlink, you can use `[member lod_max_distance]`.

File: servers/visual/visual_server_scene.cpp [hunk 2]
```diff
@@ -1936,6 +1946,33 @@ void VisualServerScene::_prepare_scene(const Transform p_cam_transform, const Ca
...
+				if (ins->prev_lod_state) {
+					lod_begin_with_hys -= ins->lod_begin_hysteresis / 2.f;
+					lod_end_with_hys   += ins->lod_end_hysteresis / 2.f;
+				} else {
+					lod_begin_with_hys += ins->lod_begin_hysteresis / 2.f;
+					lod_end_with_hys   -= ins->lod_end_hysteresis / 2.f;
+				}
+
+				if (ins->lod_begin <= 0.f) {
+					lod_begin_with_hys = -std::numeric_limits<float>::infinity();
```
Comment (akien-mga) [inline servers/visual/visual_server_scene.cpp]: We don't use `std` in Godot, you should use `Math_INF` from `core/math/math_funcs.h`.

**Comment by akien-mga** [comment]
Could you squash commits together? See [PR workflow](http://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase).

**Comment by Windfisch** [comment]
Thank you for your feedback! (And sorry for not reading these documents beforehand)
I've fixed the coding style issues and the documentation and have rebased the branch.

**Comment by Windfisch** [comment]
While playing around, I got the following godot crash which I can't reproduce, neither in release nor debug mode: To me it looks unrelated to my patch, but what's your opinion on this?
```
ERROR: generate: EditorResourcePreviewGenerator::generate needs to be overridden
   At: editor/editor_resource_preview.cpp:55.
handle_crash: Program crashed with signal 11
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[1] /usr/lib/libc.so.6(+0x368f0) [0x7ff4543d98f0] (??:0)
[2] Reference::reference() (??:0)
[3] Ref<Texture>::ref(Ref<Texture> const&) (??:0)
[4] Ref<Texture>::Ref(Ref<Texture> const&) (??:0)
[5] Button::get_icon() const (??:0)
[6] Viewport::_gui_input_event(Ref<InputEvent>) (??:0)
[7] Viewport::input(Ref<InputEvent> const&) (??:0)
[8] Viewport::_vp_input(Ref<InputEvent> const&) (??:0)
[9] MethodBind1<Ref<InputEvent> const&>::call(Object*, Variant const**, int, Variant::CallError&) (??:0)
[10] Object::call(StringName const&, Variant const**, int, Variant::CallError&) (??:0)
[11] Object::call(StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) (??:0)
[12] SceneTree::call_group_flags(unsigned int, StringName const&, StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) (??:0)
[13] SceneTree::input_event(Ref<InputEvent> const&) (??:0)
[14] InputDefault::_parse_input_event_impl(Ref<InputEvent> const&, bool) (??:0)
[15] InputDefault::parse_input_event(Ref<InputEvent> const&) (??:0)
[16] OS_X11::process_xevents() (??:0)
[17] OS_X11::run() (??:0)
[18] /home/flo/kruschkram/godot/bin/godot.x11.tools.64(main+0xd5) [0x55abf07657ba] (??:0)
[19] /usr/lib/libc.so.6(__libc_start_main+0xeb) [0x7ff4543c606b] (??:0)
[20] /home/flo/kruschkram/godot/bin/godot.x11.tools.64(_start+0x2a) [0x55abf076562a] (??:0)
-- END OF BACKTRACE --
```

**Comment by akien-mga** [comment]
Yeah the backtrace doesn't seem related to your changes. We merged a lot of patches over the last 3 days so there might be some new bugs in the master branch.

**Comment by Sslaxx** [comment]
The crash sounds like #20451.

**Comment by darrylryan** [comment]
This seems to work for me, although the shadow still gets drawn even if the mesh isnt...

**Comment by Windfisch** [comment]
yes, shadows probably need more thinking; should a high-detail-model be selected if the light's camera is close to the object, or if the main viewport's camera (which is the main camera in scenarios with multiple viewports/cameras?)? How to deal with directional lights.

But I guess it would be a good idea if only those meshes draw a shadow that are actually drawn by the camera.

**Comment by TheOnlyJoey** [comment]
I think it is acceptable that shadows are generated based on the high poly mesh to be honest, in most scenario's it creates the most expected result.

**Comment by Windfisch** [comment]
If the purpose of `lod_{min,max}_distance` is to reduce rendering workload, then this would be counter productive.
Except maybe, if it's really bound to the distance between the object and the light source. However, then the current implementation of the hysteresis needs to be changed to something that actually supports multiple viewers.

**Comment by Windfisch** [comment]
I've updated it to support hysteresis with multiple cameras as follows:
The `VisualServerScene::Camera` struct gets an added `OAHashMap<uint32, bool>` that keeps track of whether an object has been visible or not in the last frame.
the `_prepare_scene()` method got an additional parameter of type `OAHashMap *`. This can be `nullptr` (for reflection probes, which are not handled by hysteresis) or more likely, just the current camera's map.

This has (almost) no overhead when LOD is not used at all, because all the map lookups are only made for objects that have a LOD range set.

**Comment by Windfisch** [comment]
Unfortunately, objects which are LOD-culled away still seem to cast shadows; why is this?

**Comment by akien-mga** [comment]
Discussed this with @reduz on IRC:
```
16:17 <reduz> As I explained before, I am against adding this now without lod dependency
16:18 <reduz> this needs to be added as 1) lod range 2) lod depndency 3) per material lod ranges
16:18 <reduz> else its not that useful
16:18 <reduz> i was planning to add it after 3.1 is out, since i will rewrite large part of the renderer by then
```

As such, moving to the 3.2 milestone (might end up as 4.0 depending on whether or not it's bound to the Vulkan implementation).

**Pull request closed without merging**